packages/nextjs-kit/src/components/paginator.tsx

```diff
@@ -0,0 +1,234 @@
+import React, { useState, useEffect } from 'react';
+import { useRouter } from 'next/router';
+import { isNumber } from '../types';
+
+/**
+ *
+ * @param props - The props needed for the paginator component
+ * @param props.data - An array of paginator objects
+ * @param props.itemsPerPage - How many items to display per page
+ * @param props.breakpoints - Breakpoints has 3 properties: start, end, and add. Set to {} for no breakpoint.
+ * @remarks
+ *
+ * start: where to start the breakpoint
+ * end: where to end the breakpoint
+ * add: how many buttons to add when the breakpoint is clicked
+ * ***
+ * note: (`add` * x) + `start` = `end` where x is a number of clicks it takes to fill in all of the buttons
+ * For example: If there are 25 buttons and the start = 5 and end = 25, then add should be 5 or 10.
+ * ***
+ *
+ * @param props.routing If true, shallow routing will be enabled. Check the examples/pagination route to see it in action
+ * @param props.Component React Component that takes in currentItems as props and maps over them.
+ * currentItems is a subset of data, so any component that works for data will work here.
+ * @see {@link https://github.com/pantheon-systems/decoupled-kit-js/tree/canary/starters/next-drupal-starter/pages/examples/pagination/[[...page]].js} for an example implementation
+ * @returns Component with data rendered by the passed in Component and page buttons
+ */
+
+interface PaginationProps {
+  data: Array<object>;
```

Comment:
What I meant by using a generic was more for this prop, not the component. `Array<object>` doesn't give much type safety (and like we discussed in the past, it can be represented as `object[]`). Instead, we can take a generic from the user and have that be the type of the data.

```typescript
interface PaginationProps<Type> {
  data: Type[];
...
}

// then the component:
const Paginator = <Type extends object>({
  data,
  breakpoints,
  itemsPerPage,
  routing,
  Component,
}: PaginationProps) => {
...
// and as the type for current items
  const [currentItems, setCurrentItems] = useState<Type[]>([]);
```